go/processor/geth/processor.go

```diff
@@ -0,0 +1,397 @@
+	config := evmConfig(p.interpreter, p.ethereumCompatible)
+	evm := vm.NewEVM(blockContext, txContext, stateDB, chainConfig, config)
+
+	msg := transactionToMessage(transaction, blockParameters.BaseFee)
+	gasPool := new(core.GasPool).AddGas(uint64(transaction.GasLimit))
+	result, err := core.ApplyMessage(evm, msg, gasPool)
+	if err != nil {
+		if !p.ethereumCompatible && errors.Is(err, core.ErrInsufficientFunds) {
+			return tosca.Receipt{}, err
+		}
+		return tosca.Receipt{GasUsed: transaction.GasLimit}, err
+	}
+
+	createdAddress := (*tosca.Address)(&stateDB.createdContract)
+	if transaction.Recipient != nil || result.Failed() {
+		createdAddress = nil
+	}
+
+	return tosca.Receipt{
+		Success:         !result.Failed(),
+		Output:          result.ReturnData,
+		ContractAddress: createdAddress,
+		GasUsed:         tosca.Gas(result.UsedGas),
+		Logs:            stateDB.context.GetLogs(),
+	}, nil
+}
+
+func newBlockContext(blockParameters tosca.BlockParameters, context tosca.TransactionContext) vm.BlockContext {
+	canTransfer := func(stateDB vm.StateDB, address common.Address, value *uint256.Int) bool {
+		return stateDB.GetBalance(address).Cmp(value) >= 0
+	}
+
+	transfer := func(stateDB vm.StateDB, sender common.Address, recipient common.Address, value *uint256.Int) {
+		stateDB.SubBalance(sender, value, tracing.BalanceChangeTransfer)
+		stateDB.AddBalance(recipient, value, tracing.BalanceChangeTransfer)
+	}
+
+	hashFunc := func(num uint64) common.Hash {
+		return common.Hash(context.GetBlockHash(int64(num)))
+	}
+
+	return vm.BlockContext{
+		CanTransfer: canTransfer,
+		Transfer:    transfer,
+		GetHash:     hashFunc,
+		Coinbase:    common.Address(blockParameters.Coinbase),
+		GasLimit:    uint64(blockParameters.GasLimit),
+		BlockNumber: new(big.Int).SetInt64(blockParameters.BlockNumber),
+		Time:        uint64(blockParameters.Timestamp),
+		Difficulty:  big.NewInt(1),
+		BaseFee:     blockParameters.BaseFee.ToBig(),
+		BlobBaseFee: blockParameters.BlobBaseFee.ToBig(),
+		Random:      (*common.Hash)(&blockParameters.PrevRandao),
+	}
+}
+
+func blockParametersToChainConfig(blockParams tosca.BlockParameters) *params.ChainConfig {
+	chainConfig := *params.AllEthashProtocolChanges
+	chainConfig.ChainID = new(big.Int).SetBytes(blockParams.ChainID[:])
+	chainConfig.ByzantiumBlock = big.NewInt(0)
+	chainConfig.IstanbulBlock = big.NewInt(0)
+	chainConfig.BerlinBlock = big.NewInt(0)
+	chainConfig.LondonBlock = big.NewInt(0)
+	chainConfig.MergeNetsplitBlock = big.NewInt(0)
+	zeroTime := uint64(0)
+	chainConfig.ShanghaiTime = &zeroTime
+	chainConfig.CancunTime = &zeroTime
+
+	if blockParams.Revision < tosca.R13_Cancun {
+		time := uint64(blockParams.Timestamp + 1)
+		chainConfig.CancunTime = &time
+	}
+	if blockParams.Revision < tosca.R12_Shanghai {
+		time := uint64(blockParams.Timestamp + 1)
+		chainConfig.ShanghaiTime = &time
+	}
+	if blockParams.Revision < tosca.R11_Paris {
+		chainConfig.MergeNetsplitBlock = big.NewInt(blockParams.BlockNumber + 1)
```

Comment:
same here, `big.NewInt(blockParams.BlockNumber + 1)` could be reused. but neither this nor `time` are critical comments.